lib/network_ssl/network_ssl_compat.c

```diff
@@ -6,6 +6,63 @@
 
 #include "network_ssl_compat.h"
 
+#ifdef NETWORK_SSL_COMPAT_TLS_VERSION
+/**
+ * network_ssl_compat_TLS_client_method():
+ * Create a SSL_METHOD for TLS 1.2.
+ *
+ * COMPATIBILITY: provides TLS_client_method(), but only for TLS 1.2 (not
+ * based on negotiation with the server).
+ */
+const SSL_METHOD *
+network_ssl_compat_TLS_client_method()
+{
+
+	return TLSv1_2_client_method();
+}
+#endif
+
+#ifdef NETWORK_SSL_COMPAT_TLS_VERSION
+/**
+ * network_ssl_compat_CTL_set_min_proto_version(ctx, version):
+ * Set the minimum protocol version to TLS 1.2.  ${version} must be
+ * TLS_1_2_VERSION.
+ *
+ * Return 1 on success, 0 on failure.
+ *
+ * COMPATIBILITY: provides SSL_CTX_set_min_proto_version(), but only supports
+ * TLS 1.2.
+ */
+int
+network_ssl_compat_CTL_set_min_proto_version(SSL_CTX * ctx, int version)
+{
+
+	/* This the only version currently supported in this file. */
+	assert(version == TLS1_2_VERSION);
+
+	/**
+	 * We should already have a TLS 1.2 connection, because that's the
+	 * only one that network_ssl_compat_TLS_client_method() supports.  But
+	 * to be safe, we'll still disable the other options.
+	 *
+	 * Unfortunately the _set_options() function doesn't return success or
+	 * failure; instead, it returns the new bitmask after setting the
+	 * options.  So we need to AND it with the constant to verify that
+	 * it's been set.
+	 */
+	if ((SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2) & SSL_OP_NO_SSLv2) == 0)
```

Comment:
This seems unnecessary.  I don't think it's possible for SSL_CTX_set_options to fail?

Reply:
Hmm.  I can't see a serious failure path for `SSL_CTX_set_options()` in 1.0.2, but they **can** fail in 1.1.1.

### openssl 1.1.0

The function that we're aiming to replace -- `SSL_CTX_set_min_proto_version()` -- specifically says:
> These setter functions return 1 on success and 0 on failure.
> https://www.openssl.org/docs/man1.1.0/man3/SSL_CTX_set_min_proto_version.html

Can they it in practice?  Yes, at least in openssl 1.1.1 (I haven't checked 1.1.0).  For example, if openssl was compiled with `OPENSSL_NO_TLS1_2` and you attempt to set the min version to TLS 1.2, `ssl_check_allowed_versions()` will return 0, which will return 0 from:
```
    case SSL_CTRL_SET_MIN_PROTO_VERSION:
        return ssl_check_allowed_versions(larg, s->max_proto_version)
               && ssl_set_version_bound(s->ctx->method->version, (int)larg,
                                        &s->min_proto_version);
```
> https://github.com/openssl/openssl/blob/master/ssl/ssl_lib.c#L2369

Granted, you might feel safe in assuming that nobody would compile openssl without TLS 1.2.  In which case, it doesn't look like this can fail -- at least, not with the current code base.  Who knows how it might change in the future?

(granted, we're not running this compat function with openssl 1.1.0+)

### openssl 1.0.2

It looks like it can only fail if `ctx == NULL` with the current code base.

From `SSL_CTX_set_options()`, we get:

```
# define SSL_CTX_set_options(ctx,op) \
        SSL_CTX_ctrl((ctx),SSL_CTRL_OPTIONS,(op),NULL)

...
long SSL_CTX_ctrl(SSL *s, int cmd, long larg, void *parg)
{
...
    if (ctx == NULL) {
        switch (cmd) {
...
        default:
            return 0;
        }
    }
...
    switch (cmd) {
    case SSL_CTRL_OPTIONS:
        return (ctx->options |= larg);
...
}
```

But it's worth noting that the OpenSSL developers aren't adverse to failing in `SSL_CTX_ctrl()`.  In 1.0.2, the only such case is:
```
    case SSL_CTRL_SET_MAX_SEND_FRAGMENT:
        if (larg < 512 || larg > SSL3_RT_MAX_PLAIN_LENGTH)
            return 0;
        ctx->max_send_fragment = larg;
        return 1;
```

but 1.1.0 has many more `return 0`.

### openssl 1.0.3 ?

My inclination is to err on the side of caution.  It's not impossible that these functions might fail for other reasons in future openssl versions, and it would be a Very Bad Thing if the connection ended up using SSLv3 or TLS 1.0.

But yes, it would probably be ok to change this function to merely:
```
        /* This the only version currently supported in this file. */
        assert(version == TLS1_2_VERSION);

        SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2);
        SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv3);
        SSL_CTX_set_options(ctx, SSL_OP_NO_TLSv1);
        SSL_CTX_set_options(ctx, SSL_OP_NO_TLSv1_1);
```